### PR TITLE
Fix str_to_date function

### PR DESCRIPTION
**Pull request by ktat**
`STR_TO_DATE` function cannot parse "%Y%m%d".
I mentioned it in the issue #2666

### CODE REVIEW

File: sql/expression/function/str_to_date_test.go [hunk 1]
```diff
@@ -11,16 +11,24 @@ import (
...
+
 func TestStrToDate(t *testing.T) {
 	setupTimezone(t)
 
 	testCases := [...]struct {
 		name     string
 		dateStr  string
 		fmtStr   string
-		expected string
+		expected *string
```
Comment (jycor) [inline sql/expression/function/str_to_date_test.go:25]: I would just make expected of type `interface{}` or `any`

File: sql/expression/function/str_to_date_test.go [hunk 2]
```diff
@@ -11,16 +11,24 @@ import (
 	"github.com/dolthub/go-mysql-server/sql/types"
 )
 
+func ptrStr(s string) *string {
```
Comment (jycor) [inline sql/expression/function/str_to_date_test.go:14]: Remove this

File: sql/expression/function/str_to_date_test.go [hunk 3]
```diff
@@ -33,7 +41,11 @@ func TestStrToDate(t *testing.T) {
 		}
 		t.Run(tt.name, func(t *testing.T) {
 			dtime := eval(t, f, sql.NewRow(tt.dateStr, tt.fmtStr))
-			require.Equal(t, tt.expected, dtime)
+			if tt.expected != nil {
```
Comment (jycor) [inline sql/expression/function/str_to_date_test.go:44]: If `tt.expected` is of type `interface{}`, then the if check is unnecessary

**Review by jycor — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution!
The fix looks great, just some small changes to the tests, and will be happy to merge this

**Comment by ktat** [comment]
@jycor Thanks. I fiexed the code you mentioned.

**Review by jycor — APPROVED** [review]
Thanks for the changes
LGTM!

**Comment by ktat** [comment]
@jycor sorry I added additional commit. It fixes the bug of parsing "%Y%m%d%T" format.

File: sql/expression/function/str_to_date_test.go [hunk 4]
```diff
@@ -18,9 +18,17 @@ func TestStrToDate(t *testing.T) {
...
+		expected interface{}
 	}{
 		{"standard", "Dec 26, 2000 2:13:15", "%b %e, %Y %T", "2000-12-26 02:13:15"},
+		{"ymd", "20240101", "%Y%m%d", "2024-01-01"},
+		{"ymd", "2024121", "%Y%m%d", "2024-12-01"},
+		{"ymd", "20241301", "%Y%m%d", nil},
+		{"ymd", "20240001", "%Y%m%d", nil},
+		{"ymd-with-time", "2024010203:04:05", "%Y%m%d%T", "2024-01-02 03:04:05"},
+		{"ymd-with-time", "202408122:03:04", "%Y%m%d%T", "2024-08-12 02:03:04"},
+		// TODO: It shoud be nil, but returns "2024-02-31"
```
Comment (jycor) [inline sql/expression/function/str_to_date_test.go:30]: Are you going to fix this test case? I think it's ok to leave as a TODO.

**Review by jycor — APPROVED** [review]
Additional changes LGTM!
Thanks again!

File: sql/expression/function/str_to_date_test.go (hunk 4 above)
Comment (ktat) [reply sql/expression/function/str_to_date_test.go:30]: It's not a big problem. So I think it can be left as TODO.

**Pull request merged**